### DISCUSSION TITLE
Proper way to sync clients?

### DISCUSSION THREAD

**Discussion by chnamoore** [Mimetis/Dotmim.Sync#945] [Q&A] [answered]
I'm trying to make sure I'm correctly updating my client application with the latest scope setup when the application is updated. I have a xamarin app on android using an sqlite database. On the server side, an IIS web service .net core application.

Everything works fine SO FAR, but I wanted to make sure this is a recommended way to sync my clients and get them up to date.

The method I'm using -

-each time I need to update my EF model, I publish a new scope setup on the web service using the names "v1","v2", etc.
-The clients using the original scope "v1" still seem to sync with the server
-The clients updated to the new application, are only using scope "v2" using this method:

*Each time the android application starts to sync-
-First, EF migrations updates the tables on the client to the latest version when the latest app is ran.
-Next, I use the local orchestrator to get all scopes on the client. Looping through the scopes, if the name does not match the latest "v2", then it is deprovisioned locally then the scope is deleted locally.
-Next, I get the scope information for "v2" from the web service orchestrator. 
-Next, I check the local orchestrator to see if "v2" exists. If it does not, then I then provision the local client with "v2".
-Next, I sync with "v2" and use the SyncSetup that was pulled from the server.
-On the next sync, the process happens again, but nothing is deprovisioned or provisioned, because "v2" exists locally.

This works, but I'm not sure what issues this will cause if any. I'm not totally understanding on how I'm supposed to provision the clients when I need to change the database so this is what I came up with.

Aaron

**Comment by Mimetis** [comment, marked as answer]
For clients that are updating their local database using EF migrations, you know that they should use another scope.
So far, for these clients:

- 1 `EF` migrations update the tables on the client to the latest version when the latest app is starting.

Just after the EF Migrations, you already know that `V2` scope is not existing, as you just made the EF Migrations.
Of course, you can check it by getting all scopes:

- 2 `localOrchestrator` instance gets all scopes and see if a "`V2`" if exists

IF exists, I guess you are good to make a sync.

Otherwise:

- 3 `remoteOrchestrator` gets the scope information for "V2" from server side
- 4 `localOrchestrator` call `ProvisionAsync` using the remote scope retrieved on step 3
- 5 `localOrchestrator` now gets the `ScopeInfoClient` instance and update the information using the `ShadowScope` method to be sure we are going to synchronize from the good timestamp (and not from the begining)

See https://dotmimsync.readthedocs.io/Provision.html#client-1

- 6 You can sync
- 7 Optionnaly you can remove the old scope on the local client

Make sense ?

**Reply by chnamoore** [reply]
Thanks, This looks very similar to what I had mentioned except for the shadowscope park. I'll check it out today and get back to you.

**Comment by chnamoore** [comment]
If I look at the link to the docs, it says-
"// TRICKY PART
/*
    The scope v1 is new.
    **If we sync now, since v1 is new, we are going to sync all the rows from start**
    What we want is to sync from the last point we sync the old v0 scope
    That's why we are shadowing the metadata info from v0 into v1
*/"

In my case, I don't mind if the client gets all rows again- do I actually need to do the shadowScope part? And If I do not do the ShadowScope, will any updated rows still sync to the server?

Aaron

**Reply by Mimetis** [reply]
yes, DMS will make a sync from the begining.
If you don't mind overwriting existing rows locally, you're good to go and remove this step.
Make some tests and let me know if it's acceptable for you.

I would say that overwritten existing rows locally is not a big deal until you reach a point where you have _A LOT_ of rows locally (and the overwriting step could take some time)

Let me know

**Reply by chnamoore** [reply]
ok- sorry my 2nd question might of not been clear. If I do not do a shadowScope, will I lose any of the changed rows that did not get uploaded to the server yet?

**Reply by chnamoore** [reply]
Ignore my last question. I see what you are saying. It does loose the changes. One more question- what happens when a client skips a version and performs a shadowScope. For instance, can I perform a shadowScope from v0 to lets say v5 without issues?

**Reply by Mimetis** [reply]
Basically, `ShadowScope` will configure the `ScopeClientInfo` like:
- `V5` will sync all data from last time `V0` syncs, to now.

**Reply by chnamoore** [reply]
So you are saying it is ok to skip a shadowScope for v1-4?

**Reply by Mimetis** [reply]
>  I see what you are saying. It does loose the changes

Indeed, if you don't `ShadowScope()` a previous `ClientScope`, **DMS** sees this new `ClientScope` as `"IsNew==true"` and then will make a first initialize sync from server, without trying to upload anything (the first Sync is always "_kind of_" Download Only)

if you `ShadowScope()` this new `ClientScope`, then **DMS** is making a bidirectional sync, assuming that rows are already tracked.

> So you are saying it is ok to skip a shadowScope for v1-4?

Yes.
But what is the goal, in your scenario, of scope `v1`, `v2`, and `v3` ?
